File: libs/cua-driver/clients/python/tests/mcp_fixture.py
"""Deterministic executable MCP peer used by transport contract tests."""

from __future__ import annotations

import json
import sys
from typing import Any


initialized = False


def respond(request: dict[str, Any]) -> None:
    global initialized
    method = request.get("method")
    if method == "notifications/initialized":
        initialized = True
        return
    request_id = request.get("id")
    if method == "initialize":
        result: dict[str, Any] = {
            "protocolVersion": request["params"]["protocolVersion"],
            "capabilities": {},
            "serverInfo": {"name": "fixture", "version": "1"},
        }
    elif not initialized:
        payload = {
            "jsonrpc": "2.0",
            "id": request_id,
            "error": {"code": -32002, "message": "not initialized"},
        }
        print(json.dumps(payload, separators=(",", ":")), flush=True)
        return
    elif method == "test/hang":
        return
    elif method == "tools/list":
        result = {"tools": [{"name": "get_desktop_state"}]}
    elif method == "tools/call":
        params = request.get("params", {})
        result = {
            "content": [{"type": "text", "text": "fixture result"}],
            "structuredContent": {
                "name": params.get("name"),
                "arguments": params.get("arguments", {}),
            },
        }
    else:
        payload = {
            "jsonrpc": "2.0",
            "id": request_id,
            "error": {"code": -32601, "message": f"unknown method: {method}"},
        }
        print(json.dumps(payload, separators=(",", ":")), flush=True)
        return
    payload = {"jsonrpc": "2.0", "id": request_id, "result": result}
    print(json.dumps(payload, separators=(",", ":")), flush=True)


for line in sys.stdin:
    value = json.loads(line)
    if isinstance(value, dict):
        respond(value)

File: libs/cua-driver/clients/typescript/test/mcp-fixture.mjs
import { createInterface } from "node:readline"

let initialized = false
const lines = createInterface({ input: process.stdin })

lines.on("line", line => {
  const request = JSON.parse(line)
  if (request.method === "notifications/initialized") {
    initialized = true
    return
  }
  let result
  if (request.method === "initialize") {
    result = {
      protocolVersion: request.params.protocolVersion,
      capabilities: {},
      serverInfo: { name: "fixture", version: "1" },
    }
  } else if (!initialized) {
    respond(request.id, undefined, { code: -32002, message: "not initialized" })
    return
  } else if (request.method === "test/hang") {
    return
  } else if (request.method === "tools/list") {
    result = { tools: [{ name: "get_desktop_state" }] }
  } else if (request.method === "tools/call") {
    result = {
      content: [{ type: "text", text: "fixture result" }],
      structuredContent: {
        name: request.params.name,
        arguments: request.params.arguments ?? {},
      },
    }
  } else {
    respond(request.id, undefined, { code: -32601, message: `unknown method: ${request.method}` })
    return
  }
  respond(request.id, result)
})

function respond(id, result, error) {
  const response = { jsonrpc: "2.0", id }
  if (error) response.error = error
  else response.result = result
  process.stdout.write(`${JSON.stringify(response)}\n`)
}
